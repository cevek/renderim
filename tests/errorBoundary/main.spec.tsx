import {createElement, Suspense, ErrorBoundary} from 'renderim';
import {createLoadData, render} from '../base';

function FallbackWithError(props: {error: any}) {
    return props.error.message.a.b.c.d;
}
function Errored() {
    var x = {} as {a: {b: {c: JSX.Element}}};
    return x.a.b.c;
}

test('Error in fallback with parent boundary ', () => {
    const result = render(
        <div>
            Hello
            <ErrorBoundary fallback={err => err.message}>
                <ErrorBoundary
                    fallback={err => {
                        throw new Error('foo');
                    }}>
                    <Errored />
                </ErrorBoundary>
            </ErrorBoundary>
        </div>,
    );
    expect(result.tree).toMatchSnapshot('1.root');
    expect(result.getNextRestartedComponent()).toMatchSnapshot('2.Inner ErrorBoundary');
    expect(result.getNextRestartedComponent()).toMatchSnapshot('3.Outer ErrorBoundary');
    expect(result.getNextRestartedComponent()).toBeUndefined();
});

test('Error in inner fallback component with parent boundary', () => {
    const result = render(
        <div>
            Hello
            <ErrorBoundary fallback={err => err.message}>
                <ErrorBoundary fallback={err => <FallbackWithError error={err} />}>
                    <Errored />
                </ErrorBoundary>
            </ErrorBoundary>
        </div>,
    );
    expect(result.tree).toMatchSnapshot('1.root');
    expect(result.getNextRestartedComponent()).toMatchSnapshot('2.Inner ErrorBoundary');
    expect(result.getNextRestartedComponent()).toMatchSnapshot('3.Outer ErrorBoundary');
    expect(result.getNextRestartedComponent()).toBeUndefined();
});

test('Error in fallback without parent boundary', () => {
    const consoleError = jest.fn();
    console.error = consoleError;
    const result = render(
        <div>
            Hello
            <ErrorBoundary fallback={err => <FallbackWithError error={err} />}>
                <Errored />
            </ErrorBoundary>
        </div>,
    );
    expect(result.tree).toBeUndefined();
    expect(result.getNextRestartedComponent()).toMatchSnapshot('1.ErrorBoundary');
    expect(result.getNextRestartedComponent()).toBeUndefined();
    expect(consoleError).toHaveBeenCalled();
});

test('no parent boundary', () => {
    const consoleError = jest.fn();
    console.error = consoleError;
    const result = render(
        <div>
            Hello
            <Errored />
        </div>,
    );
    expect(result.tree).toBeUndefined();
    expect(result.getNextRestartedComponent()).toBeUndefined();
    expect(consoleError).toHaveBeenCalled();
});
