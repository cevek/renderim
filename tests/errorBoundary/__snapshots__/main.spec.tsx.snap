// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error in fallback with parent boundary : 1.root 1`] = `
"<div>
   Hello
   <ErrorBoundary>
      <Boundary>
         foo
      </Boundary>
   </ErrorBoundary>
</div>"
`;

exports[`Error in fallback with parent boundary : 2.Inner ErrorBoundary 1`] = `"<ErrorBoundary/>"`;

exports[`Error in fallback with parent boundary : 3.Outer ErrorBoundary 1`] = `
"<ErrorBoundary>
   <Boundary>
      foo
   </Boundary>
</ErrorBoundary>"
`;

exports[`Error in fallback without parent boundary: 1.ErrorBoundary 1`] = `
"<ErrorBoundary>
   <Boundary>
      <FallbackWithError/>
   </Boundary>
</ErrorBoundary>"
`;

exports[`Error in inner fallback component with parent boundary: 1.root 1`] = `
"<div>
   Hello
   <ErrorBoundary>
      <Boundary>
         Cannot read property 'b' of undefined
      </Boundary>
   </ErrorBoundary>
</div>"
`;

exports[`Error in inner fallback component with parent boundary: 2.Inner ErrorBoundary 1`] = `
"<ErrorBoundary>
   <Boundary>
      <FallbackWithError/>
   </Boundary>
</ErrorBoundary>"
`;

exports[`Error in inner fallback component with parent boundary: 3.Outer ErrorBoundary 1`] = `
"<ErrorBoundary>
   <Boundary>
      Cannot read property 'b' of undefined
   </Boundary>
</ErrorBoundary>"
`;
